README.md
[![Flutter Community: flutter_whatsnew](https://fluttercommunity.dev/_github/header/flutter_whatsnew)](https://github.com/fluttercommunity/community)

# flutter_whatsnew

A Flutter Plugin to Show a Whats New page. 

Online Demo: https://fluttercommunity.github.io/flutter_whatsnew/

## Usage

To use this plugin, add `flutter_whatsnew` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).

``` dart
// Import package
import 'package:flutter_whatsnew/flutter_whatsnew.dart';
```

![changelog](https://github.com/fluttercommunity/flutter_whatsnew/blob/master/doc/screenshots/changes.png)

If using `WhatsNewPage.changelog` and you want it to pickup the `CHANGELOG.md' add it to the Flutter asset directory, otherwise manually pass in the .md file with the changes. This uses the full markdown previewer.

``` yaml
  assets:
    - CHANGELOG.md
```

![changelog](https://github.com/fluttercommunity/flutter_whatsnew/blob/master/doc/screenshots/changelog.png)

## Example

``` dart
import 'package:flutter/material.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter_whatsnew/flutter_whatsnew.dart';

void main() {
  runApp(MaterialApp(home: ShowWhatsNew()));
}

class ShowWhatsNew extends StatelessWidget {
  final double textScaleFactor = 1.0;

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      body: SafeArea(
          child: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: <Widget>[
            RaisedButton(
              child: Text("Show Changelog"),
              onPressed: () {
                Navigator.push(
                  context,
                  MaterialPageRoute(
                    builder: (context) => WhatsNewPage.changelog(
                          title: Text(
                            "What's New",
                            textScaleFactor: textScaleFactor,
                            textAlign: TextAlign.center,
                            style: const TextStyle(
                              // Text Style Needed to Look like iOS 11
                              fontSize: 22.0,
                              fontWeight: FontWeight.bold,
                            ),
                          ),
                          buttonText: Text(
                            'Continue',
                            textScaleFactor: textScaleFactor,
                            style: const TextStyle(color: Colors.white),
                          ),
                        ),
                    fullscreenDialog: true,
                  ),
                );
              },
            ),
            Container(height: 50.0),
            RaisedButton(
              child: Text("Show Changes"),
              onPressed: () {
                Navigator.push(
                  context,
                  MaterialPageRoute(
                    builder: (context) => WhatsNewPage(
                          title: Text(
                            "What's New",
                            textScaleFactor: textScaleFactor,
                            textAlign: TextAlign.center,
                            style: const TextStyle(
                              // Text Style Needed to Look like iOS 11
                              fontSize: 22.0,
                              fontWeight: FontWeight.bold,
                            ),
                          ),
                          buttonText: Text(
                            'Continue',
                            textScaleFactor: textScaleFactor,
                            style: const TextStyle(color: Colors.white),
                          ),
                          // Create a List of WhatsNewItem for use in the Whats New Page
                          // Create as many as you need, it will be scrollable
                          items: <ListTile>[
                            ListTile(
                              leading: const Icon(Icons.color_lens),
                              title: Text(
                                'Dark Theme',
                                textScaleFactor: textScaleFactor,
                              ), //Title is the only Required Item
                              subtitle: Text(
                                'Black and grey theme (Tap to Change)',
                                textScaleFactor: textScaleFactor,
                              ),
                              onTap: () {
                                // You Can Navigate to Locations in the App
                                Navigator.of(context).pushNamed("/settings");
                              },
                            ),
                            ListTile(
                              leading: const Icon(Icons.map),
                              title: Text(
                                'Google Maps',
                                textScaleFactor: textScaleFactor,
                              ),
                              subtitle: Text(
                                'Open Address Links in Google Maps instead of Apple Maps (Tap to Change)',
                                textScaleFactor: textScaleFactor,
                              ),
                              onTap: () {
                                // You Can Navigate to Locations in the App
                                Navigator.of(context).pushNamed("/settings");
                              },
                            ),
                            ListTile(
                              leading: const Icon(Icons.notifications_active),
                              title: Text(
                                'Push Notifications',
                                textScaleFactor: textScaleFactor,
                              ),
                              subtitle: Text(
                                'Stay tuned for important information that can be pushed to you',
                                textScaleFactor: textScaleFactor,
                              ),
                              onTap: () {
                                WhatsNewPage.showDetailPopUp(
                                  context,
                                  'Info',
                                  "You can turn off push notifications any time in your application settings.",
                                );
                              },
                            ),
                          ], //Required
                        ),
                    fullscreenDialog: true,
                  ),
                );
              },
            ),
          ],
        ),
      )),
    );
  }
}

```


CHANGELOG.md
## 1.0.4

* Update dependencies
* Fix Flutter/Dart version

## 1.0.3

* Adding Web Support

## 1.0.2

* Adding Path Option
* Removing Native Widgets

## 1.0.1

* Bumping Version

## 1.0.0

* Addign Cupertino UI

## 0.3.0 - 04.06.2019

* Making Example Desktop Aware

## 0.2.1 - 04.05.2019

* `tb2761` Added Button Colors

## [0.2.0] - Breaking Changes

* Updated Example
* Added Ability to Read from Changelog
* Removed Method WhatsNewPage.show()
* Adding `flutter_markdown`

## [0.1.0] - Breaking Change

* Updated Example
* Removed `get_version`
* Plugin is now pure dart
* Added Method WhatsNewPage.show()

## [0.0.7] - New Button Action

* Updated Example

## [0.0.6] - New Button Action

* Added Optional Callback for Button Press.
* By Defualt it will Push to the Passed Widget.
* Updating Dart Dependency

## [0.0.5] - Updating for Flutter

* Increased button width on android.
* Updated native_widgets 0.1.0

## [0.0.4] - Bug Fixes

* Fixed Loading Bug.

## [0.0.3] - Bug Fixes

* Updating Startup Behavior.

## [0.0.2] - Bug Fixes

* Updating Startup Behavior.

## [0.0.1] - Whats New Widget

* Added Components, Version Check and Example.
* Added `native_widgets`


lib/flutter_whatsnew.dart
library flutter_whatsnew;

import 'dart:io';

import 'package:flutter/cupertino.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_markdown/flutter_markdown.dart';

class WhatsNewPage extends StatelessWidget {
  final Widget title;
  final Widget buttonText;
  final List<ListTile> items;
  final VoidCallback onButtonPressed;
  final bool changelog;
  final String changes;
  final Color backgroundColor;
  final Color buttonColor;
  final String path;

  const WhatsNewPage({
    @required this.items,
    @required this.title,
    @required this.buttonText,
    this.onButtonPressed,
    this.backgroundColor,
    this.buttonColor,
  })  : changelog = false,
        changes = null,
        path = null;

  const WhatsNewPage.changelog({
    @required this.title,
    @required this.buttonText,
    this.onButtonPressed,
    this.changes,
    this.backgroundColor,
    this.buttonColor,
    this.path,
  })  : changelog = true,
        items = null;

  static void showDetailPopUp(
      BuildContext context, String title, String detail) async {
    void showDemoDialog<T>({BuildContext context, Widget child}) {
      showDialog<T>(
        context: context,
        barrierDismissible: false,
        builder: (BuildContext context) => child,
      );
    }

    return showDemoDialog<Null>(
      context: context,
      child: AlertDialog(
        title: Text(title),
        content: Text(detail),
        actions: <Widget>[
          TextButton(
            child: Text('OK'),
            onPressed: () {
              Navigator.pop(context);
            },
          ),
        ],
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    print("Changelog: $changelog");
    if (!kIsWeb && Platform.isIOS) {
      return _buildIOS(context);
    }

    return _buildAndroid(context);
  }

  Widget _buildAndroid(BuildContext context) {
    if (changelog) {
      return Scaffold(
        backgroundColor:
            backgroundColor ?? Theme.of(context).scaffoldBackgroundColor,
        body: SafeArea(
          child: Stack(
            fit: StackFit.loose,
            children: <Widget>[
              Positioned(
                top: 10.0,
                left: 0.0,
                right: 0.0,
                child: title,
              ),
              Positioned(
                left: 0.0,
                right: 0.0,
                top: 50.0,
                bottom: 80.0,
                child: ChangeLogView(changes: changes, path: path),
              ),
              Positioned(
                bottom: 5.0,
                right: 10.0,
                left: 10.0,
                child: ListTile(
                  title: ElevatedButton(
                    child: buttonText,
                    style: ButtonStyle(
                        backgroundColor: MaterialStateProperty.all(
                            buttonColor ?? Colors.blue)),
                    onPressed: onButtonPressed != null
                        ? onButtonPressed
                        : () {
                            Navigator.pop(context);
                          },
                  ),
                ),
              ),
            ],
          ),
        ),
      );
    }
    return Scaffold(
      backgroundColor:
          backgroundColor ?? Theme.of(context).scaffoldBackgroundColor,
      body: SafeArea(
        child: Stack(
          fit: StackFit.loose,
          children: <Widget>[
            Positioned(
              top: 10.0,
              left: 0.0,
              right: 0.0,
              child: title,
            ),
            Positioned(
              left: 0.0,
              right: 0.0,
              top: 50.0,
              bottom: 80.0,
              child: ListView(
                children: items
                    .map(
                      (ListTile item) => ListTile(
                        title: item.title,
                        subtitle: item.subtitle,
                        leading: item.leading,
                        trailing: item.trailing,
                        onTap: item.onTap,
                        onLongPress: item.onLongPress,
                      ),
                    )
                    .toList(),
              ),
            ),
            Positioned(
              bottom: 5.0,
              right: 10.0,
              left: 10.0,
              child: ListTile(
                title: ElevatedButton(
                  child: buttonText,
                  style: ButtonStyle(
                    backgroundColor:
                        MaterialStateProperty.all(buttonColor ?? Colors.blue),
                  ),
                  onPressed: onButtonPressed != null
                      ? onButtonPressed
                      : () => Navigator.pop(context),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildIOS(BuildContext context) {
    Widget child;
    if (changelog) {
      child = ChangeLogView(changes: changes, path: path);
    } else {
      child = Material(
        child: ListView(
          children: items,
        ),
      );
    }
    return CupertinoPageScaffold(
      navigationBar: CupertinoNavigationBar(
        middle: title,
        automaticallyImplyMiddle: false,
        trailing: CupertinoButton(
          child: buttonText,
          onPressed: onButtonPressed != null
              ? onButtonPressed
              : () {
                  Navigator.pop(context);
                },
        ),
      ),
      child: SafeArea(
        child: child,
      ),
    );
  }
}

class ChangeLogView extends StatefulWidget {
  const ChangeLogView({this.changes, this.path});
  final String changes;
  final String path;
  @override
  _ChangeLogViewState createState() => _ChangeLogViewState();
}

class _ChangeLogViewState extends State<ChangeLogView> {
  String _changelog;

  @override
  void initState() {
    if (widget?.changes == null) {
      rootBundle.loadString(widget?.path ?? "CHANGELOG.md").then((data) {
        setState(() {
          _changelog = data;
        });
      });
    } else {
      setState(() {
        _changelog = widget.changes;
      });
    }
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    if (_changelog == null) {
      return CircularProgressIndicator();
    }
    return Markdown(data: _changelog);
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_whatsnew/flutter_whatsnew.dart';

import 'dart:io' show Platform;
import 'package:flutter/foundation.dart';

// The existing imports
// !! Keep your existing impots here !!

/// main is entry point of Flutter application
void main() {
  // Desktop platforms aren't a valid platform.
  if (!kIsWeb) _setTargetPlatformForDesktop();
  return runApp(MyApp());
}

/// If the current platform is desktop, override the default platform to
/// a supported platform (iOS for macOS, Android for Linux and Windows).
/// Otherwise, do nothing.
void _setTargetPlatformForDesktop() {
  TargetPlatform targetPlatform;
  if (Platform.isMacOS) {
    targetPlatform = TargetPlatform.iOS;
  } else if (Platform.isLinux || Platform.isWindows) {
    targetPlatform = TargetPlatform.android;
  }
  if (targetPlatform != null) {
    debugDefaultTargetPlatformOverride = targetPlatform;
  }
}

class MyApp extends StatelessWidget {
  final double textScaleFactor = 1.0;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: HomeScreen(textScaleFactor: textScaleFactor),
    );
  }
}

class HomeScreen extends StatelessWidget {
  const HomeScreen({
    Key key,
    @required this.textScaleFactor,
  }) : super(key: key);

  final double textScaleFactor;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: <Widget>[
              ElevatedButton(
                child: Text("Show Changelog"),
                onPressed: () {
                  Navigator.push(
                    context,
                    MaterialPageRoute(
                      builder: (context) => WhatsNewPage.changelog(
                        title: Text(
                          "What's New",
                          textScaleFactor: textScaleFactor,
                          textAlign: TextAlign.center,
                          style: const TextStyle(
                            // Text Style Needed to Look like iOS 11
                            fontSize: 22.0,
                            fontWeight: FontWeight.bold,
                          ),
                        ),
                        buttonText: Text(
                          'Continue',
                          textScaleFactor: textScaleFactor,
                          style: const TextStyle(color: Colors.white),
                        ),
                      ),
                      fullscreenDialog: true,
                    ),
                  );
                },
              ),
              Container(height: 50.0),
              ElevatedButton(
                child: Text("Show Changes"),
                onPressed: () {
                  Navigator.push(
                    context,
                    MaterialPageRoute(
                      builder: (context) => WhatsNewPage(
                        title: Text(
                          "What's New",
                          textScaleFactor: textScaleFactor,
                          textAlign: TextAlign.center,
                          style: const TextStyle(
                            // Text Style Needed to Look like iOS 11
                            fontSize: 22.0,
                            fontWeight: FontWeight.bold,
                          ),
                        ),
                        buttonText: Text(
                          'Continue',
                          textScaleFactor: textScaleFactor,
                          style: const TextStyle(color: Colors.white),
                        ),
                        // Create a List of WhatsNewItem for use in the Whats New Page
                        // Create as many as you need, it will be scrollable
                        items: <ListTile>[
                          ListTile(
                            leading: const Icon(Icons.color_lens),
                            title: Text(
                              'Dark Theme',
                              textScaleFactor: textScaleFactor,
                            ), //Title is the only Required Item
                            subtitle: Text(
                              'Black and grey theme (Tap to Change)',
                              textScaleFactor: textScaleFactor,
                            ),
                            onTap: () {
                              // You Can Navigate to Locations in the App
                              Navigator.of(context).pushNamed("/settings");
                            },
                          ),
                          ListTile(
                            leading: const Icon(Icons.map),
                            title: Text(
                              'Google Maps',
                              textScaleFactor: textScaleFactor,
                            ),
                            subtitle: Text(
                              'Open Address Links in Google Maps instead of Apple Maps (Tap to Change)',
                              textScaleFactor: textScaleFactor,
                            ),
                            onTap: () {
                              // You Can Navigate to Locations in the App
                              Navigator.of(context).pushNamed("/settings");
                            },
                          ),
                          ListTile(
                            leading: const Icon(Icons.notifications_active),
                            title: Text(
                              'Push Notifications',
                              textScaleFactor: textScaleFactor,
                            ),
                            subtitle: Text(
                              'Stay tuned for important information that can be pushed to you',
                              textScaleFactor: textScaleFactor,
                            ),
                            onTap: () {
                              WhatsNewPage.showDetailPopUp(
                                context,
                                'Info',
                                "You can turn off push notifications any time in your application settings.",
                              );
                            },
                          ),
                        ], //Required
                      ),
                      fullscreenDialog: true,
                    ),
                  );
                },
              ),
            ],
          ),
        ),
      ),
    );
  }
}
